plugins/newrelic/src/api/index.ts

```diff
@@ -71,39 +76,105 @@ export interface NewRelicApi {
+  private readonly fetchApi: FetchApi;
   private readonly proxyPathBase: string;
+  private baseUrl: string;
 
   constructor(options: Options) {
     this.discoveryApi = options.discoveryApi;
+    this.fetchApi = options.fetchApi;
     this.proxyPathBase = options.proxyPathBase ?? DEFAULT_PROXY_PATH_BASE;
+    this.baseUrl = '';
   }
 
   async getApplications(): Promise<NewRelicApplications> {
-    const url = await this.getApiUrl('apm', 'applications.json');
-    const response = await fetch(url);
-    let responseJson;
+    if (!this.baseUrl) {
+      const proxyUrl = await this.discoveryApi.getBaseUrl('proxy');
+      this.baseUrl = `${proxyUrl}${this.proxyPathBase}/apm/api/applications.json`;
+    }
+
+    let targetUrl = this.baseUrl;
+    let hasNextPage = true;
 
     try {
-      responseJson = await response.json();
+      let applications: NewRelicApplication[] = [];
+
+      do {
+        const { hasMoreApplications, pagination, applicationsFromReadPage } =
+          await this.fetchNewRelic(targetUrl);
+
+        hasNextPage = hasMoreApplications;
+        targetUrl = hasNextPage ? pagination!.nextPageLink : '';
+        applications = applications.concat(applicationsFromReadPage);
+      } while (hasNextPage);
+
+      return { applications };
     } catch (e) {
-      responseJson = { applications: [] };
+      return { applications: [] };
     }
+  }
+
+  private async fetchNewRelic(
+    targetUrl: string,
+  ): Promise<NewRelicPageReadResult> {
+    const response = await this.fetchApi.fetch(targetUrl);
+    const responseJson = await response.json();
 
-    if (response.status !== 200) {
+    if (!response.ok) {
       throw new Error(
         `Error communicating with New Relic: ${
           responseJson?.error?.title || response.statusText
         }`,
       );
     }
 
-    return responseJson;
+    const readResponse = responseJson as NewRelicApplications;
+    const linkHeader =
+      response.headers.get('Link') || response.headers.get('link');
+    const pagination = this.parseLinkHeader(linkHeader);
+
+    return {
+      hasMoreApplications: !!(pagination && pagination.relation === 'next'),
+      pagination,
+      applicationsFromReadPage: readResponse.applications,
+    };
   }
 
-  private async getApiUrl(product: string, path: string) {
-    const proxyUrl = await this.discoveryApi.getBaseUrl('proxy');
-    return `${proxyUrl}${this.proxyPathBase}/${product}/api/${path}`;
+  private parseLinkHeader(
+    linkHeader: string | null,
+  ): PaginationInformation | undefined {
+    if (!linkHeader) {
+      return undefined;
+    }
+
+    const nextRelevantLink = linkHeader.replaceAll(' ', '').split(',')[0];
+    const linkParts = nextRelevantLink.match(/^<.+(\?page=.+)>;rel="(.+)"$/);
+    const nextPageNumber = linkParts?.[1];
+    const relation = linkParts?.[2];
+
+    const nextPageLink = `${this.baseUrl}${nextPageNumber}`;
+    const isValidLink = !!(!!linkParts && nextPageNumber && relation);
```

Comment:
```suggestion
    const isValidLink = !!(linkParts && nextPageNumber && relation);
```

no need for the inner one